common/kvstore/src/main/java/org/apache/spark/util/kvstore/KVStoreSnapshotter.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util.kvstore;
+
+import com.google.common.io.ByteStreams;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+
+public class KVStoreSnapshotter {
+  private static final int MARKER_END_OF_TYPE = -2;
+  private static final int MARKER_END_OF_FILE = -1;
+
+  private final KVStoreSerializer serializer;
+
+  public KVStoreSnapshotter(KVStoreSerializer serializer) {
+    this.serializer = serializer;
+  }
+
+  public void dump(KVStore store, File snapshotFile) throws Exception {
+    DataOutputStream output = new DataOutputStream(new FileOutputStream(snapshotFile));
+
+    // store metadata if it exists
+    Class<?> metadataType = store.metadataType();
+    if (metadataType != null) {
+      writeClassName(metadataType, output);
+      Object metadata = store.getMetadata(metadataType);
+      writeObject(metadata, output);
+      writeEndOfType(output);
+    } else {
+      writeEndOfType(output);
+    }
+
+    Set<Class<?>> types = store.types();
+    for (Class<?> clazz : types) {
+      writeClassName(clazz, output);
+
+      KVStoreView<?> view = store.view(clazz);
```

Comment:
I'm wondering will there be a type with empty objects in the KVStore. Normally, it seems impossible.

Reply:
I'm not sure any implementations allow type with empty objects.

I see what you say - if there's some implementation allow the case like types() returning Class<A> but view(A.class) contains nothing, we don't provide the way to only add type to KVStore. We may want to be clear in the interface javadoc that "type with empty objects are ignored while recovering, so implementations should not rely on this", as thinking theoretically, but I'm afraid I might be over-thinking.